// @azure/opentelemetry-instrumentation-azure-sdk/dist/commonjs/spanWrapper.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenTelemetrySpanWrapper = void 0;
const api_1 = require("@opentelemetry/api");
const core_1 = require("@opentelemetry/core");
class OpenTelemetrySpanWrapper {
    constructor(span) {
        this._span = span;
    }
    setStatus(status) {
        if (status.status === "error") {
            if (status.error) {
                this._span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: status.error.toString() });
                this.recordException(status.error);
            }
            else {
                this._span.setStatus({ code: api_1.SpanStatusCode.ERROR });
            }
        }
        else if (status.status === "success") {
            this._span.setStatus({ code: api_1.SpanStatusCode.OK });
        }
    }
    setAttribute(name, value) {
        if (value !== null && value !== undefined && (0, core_1.isAttributeValue)(value)) {
            this._span.setAttribute(name, value);
        }
    }
    end() {
        this._span.end();
    }
    recordException(exception) {
        this._span.recordException(exception);
    }
    isRecording() {
        return this._span.isRecording();
    }
    addEvent(name, options = {}) {
        this._span.addEvent(name, (0, core_1.sanitizeAttributes)(options.attributes), options.startTime);
    }
    /**
     * Allows getting the wrapped span as needed.
     * @internal
     *
     * @returns The underlying span
     */
    unwrap() {
        return this._span;
    }
}
exports.OpenTelemetrySpanWrapper = OpenTelemetrySpanWrapper;
